Code/Framework/AzCore/AzCore/DOM/DomValue.h

```diff
@@ -0,0 +1,395 @@
+        using ShortStringType = AZStd::fixed_string<ShortStringSize>;
+        using SharedStringContainer = AZStd::vector<char>;
+        using SharedStringType = AZStd::shared_ptr<const SharedStringContainer>;
+        using OpaqueStorageType = AZStd::shared_ptr<AZStd::any>;
+
+        //! The internal storage type for Value.
+        //! These types do not correspond one-to-one with the Value's external Type as there may be multiple storage classes
+        //! for the same type in some instances, such as string storage.
+        using ValueType = AZStd::variant<
+            // Null
+            AZStd::monostate,
+            // Int64
+            int64_t,
+            // Uint64
+            uint64_t,
+            // Double
+            double,
+            // Bool
+            bool,
+            // StringType
+            AZStd::string_view,
+            SharedStringType,
+            ShortStringType,
+            // Object
+            ObjectPtr,
+            // Array
+            ArrayPtr,
+            // Node
+            NodePtr,
+            // Opaque
+            OpaqueStorageType>;
+
+        // Constructors...
+        Value() = default;
+        Value(const Value&);
+        Value(Value&&) noexcept;
+        Value(AZStd::string_view stringView, bool copy);
+        Value(SharedStringType sharedString);
+
+        Value(int8_t value);
+        Value(uint8_t value);
+        Value(int16_t value);
+        Value(uint16_t value);
+        Value(int32_t value);
+        Value(uint32_t value);
+        Value(int64_t value);
+        Value(uint64_t value);
+        Value(float value);
+        Value(double value);
+        Value(bool value);
+
+        explicit Value(Type type);
+
+        static Value FromOpaqueValue(const AZStd::any& value);
+
+        // Equality / comparison / swap...
+        Value& operator=(const Value&);
+        Value& operator=(Value&&) noexcept;
+
+        bool operator==(const Value& rhs) const;
+        bool operator!=(const Value& rhs) const;
+
+        void Swap(Value& other) noexcept;
+
+        // Type info...
+        Type GetType() const;
+        bool IsNull() const;
+        bool IsFalse() const;
+        bool IsTrue() const;
+        bool IsBool() const;
+        bool IsNode() const;
+        bool IsObject() const;
+        bool IsArray() const;
+        bool IsOpaqueValue() const;
+        bool IsNumber() const;
+        bool IsInt() const;
+        bool IsUint() const;
+        bool IsDouble() const;
+        bool IsString() const;
+
+        // Object API (also used by Node)...
+        Value& SetObject();
+        size_t MemberCount() const;
+        size_t MemberCapacity() const;
+        bool ObjectEmpty() const;
+
+        Value& operator[](KeyType name);
+        const Value& operator[](KeyType name) const;
+        Value& operator[](AZStd::string_view name);
+        const Value& operator[](AZStd::string_view name) const;
```

Comment:
With the API being very explicit what it operates on (even to the point that non-const function are explicitly called `Mutable` when there's a const overload, which seems a bit too much "C-style" imho), using `operator[]` seems uncharacteristic as it's not immediately clear this applies to objects only. They're also a bit problematic imho for `const` versions because no new value can be created, so a check for existence needs to happen before that first, but with `FindMember` the value can also be returned, while `operator[]` would need to do another lookup. Personally I would not add these operators and be more explicit such as `const Value* GetMember(KeyType name) const`.

Reply:
Bit of explanation on the "Mutable" API - it wasn't originally like that, but what I found when doing tuning is that you need to be careful using non-const APIs here because they can trigger an expensive copy-on-write operation. `operator[]` is much less explicit but I'm finding the in-practice benefits to chaining the operator helpful, things like `value["obj"]["arr"][2]` compose much better in practice then a bunch of method chaining.